tokio/src/time/driver/entry.rs

```diff
@@ -1,362 +1,616 @@
+    ) -> NonNull<linked_list::Pointers<Self::Target>> {
+        unsafe { NonNull::new(target.as_ref().driver_state.0.pointers.get()).unwrap() }
     }
+}
 
-    pub(crate) fn cancel(entry: &Arc<Entry>) {
-        let state = entry.state.fetch_or(ELAPSED, SeqCst);
+// ===== impl Entry =====
 
-        if is_elapsed(state) {
-            // Nothing more to do
-            return;
-        }
+impl TimerEntry {
+    pub(crate) fn new(handle: &Handle, deadline: Instant) -> Self {
+        let driver = handle.clone();
 
-        // If registered with a timer instance, try to upgrade the Arc.
-        let inner = match entry.upgrade_inner() {
-            Some(inner) => inner,
-            None => return,
-        };
+        Self {
+            driver,
+            inner: StdUnsafeCell::new(TimerShared::new()),
+            initial_deadline: Some(deadline),
+        }
+    }
 
-        let _ = inner.queue(entry);
+    fn inner(&self) -> &TimerShared {
+        unsafe { &*self.inner.get() }
     }
 
-    pub(crate) fn poll_elapsed(&self, cx: &mut task::Context<'_>) -> Poll<Result<(), Error>> {
-        let mut curr = self.state.load(SeqCst);
+    pub(crate) fn is_elapsed(&self) -> bool {
+        !self.inner().state.is_registered()
+    }
 
-        if is_elapsed(curr) {
-            return Poll::Ready(if curr == ERROR {
-                Err(Error::from_u8(self.error.load(SeqCst)))
-            } else {
-                Ok(())
-            });
-        }
+    /// Cancels and deregisters the timer. This operation is irreversible.
+    pub(crate) fn cancel(self: Pin<&mut Self>) {
+        // We need to perform an acq/rel fence with the driver thread, and the
+        // simplest way to do so is to grab the driver lock.
+        //
+        // Why is this necessary? We're about to release this timer's memory for
+        // some other non-timer use. However, we've been doing a bunch of
+        // relaxed (or even non-atomic) writes from the driver thread, and we'll
+        // be doing more from _this thread_ (as this memory is interpreted as
+        // something else).
+        //
+        // It is critical to ensure that, from the point of view of the driver,
+        // those future non-timer writes happen-after the timer is fully fired,
+        // and from the purpose of this thread, the driver's writes all
+        // happen-before we drop the timer. This in turn requires us to perform
+        // an acquire-release barrier in _both_ directions between the driver
+        // and dropping thread.
+        //
+        // This lock acquisition serves this purpose. All of the driver
+        // manipulations happen with the lock held, so we can just take the lock
+        // and be sure that this drop happens-after everything the driver did so
+        // far and happens-before everything the driver does in the future.
+        // While we have the lock held, we also go ahead and deregister the
+        // entry if necessary.
```

Comment:
Where does the lock acquisition you mention in the comment actually occur? It looks like that happens inside `clear_entry`, it might be good if that was clearer in the comment.